jddd-architecture/jddd-architecture-onion/src/main/java/org/jddd/architecture/onion/DomainServiceRing.java

```diff
@@ -0,0 +1,40 @@
+/*
+ * Copyright 2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jddd.architecture.onion;
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+/**
+ * Identifies the {@link DomainServiceRing} in an onion architecture. The domain service ring defines the interfaces
```

Comment:
I would not only focus on repositories for the domain service ring. In my understanding, the domain services ring would also contain "domain services" (according to the "services" internal building block of Evans) and thus not only interfaces. But that might actually also be a honey pot / misnomer as one could also argue that services belong to the domain model (as they contain domain logic that does not naturally fit onto an aggregate).

Reply:
This part is indeed difficult in the "4-ring" onion architecture.

In the domain model ring, we have the data + logic that operates directly on that data, i.e. everything which belongs naturally to the aggregate. But if that's the case and we assume that the logic also needs supporting business logic, the supporting services must also be in the domain model ring, otherwise, the model would have to depend on the service ring, which is not allowed. I think this 4-ring style will only work in 3 cases:

1. Theroretical cases were the business logic is self-consistent, i.e. there is no additional service code required which cannot be implemented inside aggregates.
2. The domain model ring consists only of the data and minimal business logic in order to keep the state of an aggregate consistent. Everything else would need to be implemented in the domain service ring.
3. The domain service ring consists only of repository interfaces, which will be called by the orchestrating application layer after the business logic was executed in the domain model layer.

As Palermo stated:
> The first layer around the Domain Model is typically where we would find interfaces that provide object saving and retrieving behavior, called repository interfaces.  The object saving behavior is not in the application core, however, because it typically involves a database.  Only the interface is in the application core.